src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -175,11 +175,261 @@ private BoundExpression BindSwitchExpression(SwitchExpressionSyntax node, Bindin
+                    var errorCode = node.IsKind(SyntaxKind.ListPatternClause)
+                        ? ErrorCode.ERR_UnsupportedTypeForListPattern
+                        : ErrorCode.ERR_UnsupportedTypeForLengthPattern;
+                    Error(diagnostics, errorCode, node, inputType.ToDisplayString());
+                }
+                hasErrors = true;
+            }
+            lookupResult.Free();
+        }
+
+        private BoundListPatternClause BindListPatternClause(
+            PropertyPatternClauseSyntax node,
+            TypeSymbol inputType,
+            bool permitDesignations,
+            ref PropertySymbol? lengthProperty,
+            ref bool hasErrors,
+            BindingDiagnosticBag diagnostics)
+        {
+            Debug.Assert(node.IsKind(SyntaxKind.ListPatternClause));
+
+            if (lengthProperty is null)
+            {
+                LookupLengthOrCountProperty(node, inputType, out lengthProperty, ref hasErrors, diagnostics);
+            }
+
+            TypeSymbol elementType;
+            BoundIndexerAccess? indexerAccess = null;
+            PropertySymbol? indexerSymbol = null;
+            if (inputType.IsSZArray())
+            {
+                elementType = ((ArrayTypeSymbol)inputType).ElementType;
+            }
+            else if (TryFindIndexerOrIndexerPattern(node, inputType, argIsIndex: true, out indexerAccess, out Symbol? patternSymbol, diagnostics))
+            {
+                if (patternSymbol is PropertySymbol indexer)
+                {
+                    indexerSymbol = indexer;
+                    elementType = indexer.Type;
+                }
+                else
+                {
+                    Debug.Assert(indexerAccess is not null);
+                    elementType = indexerAccess.Type;
+                }
+            }
+            else
+            {
+                if (!hasErrors && !inputType.IsErrorType())
+                    Error(diagnostics, ErrorCode.ERR_UnsupportedTypeForListPattern, node, inputType);
+                hasErrors = true;
+                elementType = CreateErrorType();
+            }
+
+            ImmutableArray<BoundPattern> subpatterns = BindListPatternSubpatterns(node.Subpatterns, inputType, elementType, permitDesignations, ref hasErrors, out bool sawSlice, diagnostics);
+            return new BoundListPatternClause(node, subpatterns, sawSlice, indexerAccess, indexerSymbol, hasErrors);
+        }
+
+        private bool TryFindIndexerOrIndexerPattern(
+            SyntaxNode syntax, TypeSymbol receiverType, bool argIsIndex,
+            out BoundIndexerAccess? indexerAccess,
+            out Symbol? patternSymbol,
+            BindingDiagnosticBag diagnostics)
+        {
+            indexerAccess = null;
+            patternSymbol = null;
+            TypeSymbol argType = Compilation.GetWellKnownType(argIsIndex ? WellKnownType.System_Index : WellKnownType.System_Range);
+            if (argType.IsErrorType())
```

Comment:
Maybe my suggestion to move this before getting an instance of `LookupResult` messed something up.

I'm getting unexpected errors with a scenario that worked a few commits earlier, because the Index/Range types aren't available.

```
        [Theory]
        [InlineData("{ 42, ..var unused }")]
        public void ListPattern_LengthAndIndexAndSliceCall(string pattern)
        {
            var source = $@"
System.Console.Write(new C() is {pattern});

public class C
{{
    public int Length {{ get {{ System.Console.Write(""Length ""); return 1; }} }}
    public int this[int i]  {{ get {{ System.Console.Write(""Index ""); return 42; }} }}
    public int Slice(int i, int j) {{ System.Console.Write(""Slice ""); return 0; }}
}}
";
            var compilation = CreateCompilation(source, parseOptions: TestOptions.RegularWithListPatterns);
            // unexpected errors
            compilation.VerifyDiagnostics(
                    // (2,33): error CS9200: List patterns may not be used for a value of type 'C'.
                    // System.Console.Write(new C() is { 42, ..var unused });
                    Diagnostic(ErrorCode.ERR_UnsupportedTypeForListPattern, "{ 42, ..var unused }").WithArguments("C").WithLocation(2, 33),
                    // (2,39): error CS9201: Slice patterns may not be used for a value of type 'C'.
                    // System.Console.Write(new C() is { 42, ..var unused });
                    Diagnostic(ErrorCode.ERR_UnsupportedTypeForSlicePattern, "..var unused").WithArguments("C").WithLocation(2, 39)
                    );
            //CompileAndVerify(compilation, expectedOutput: "Length Index Slice True");
        }
```